src/interpreter/mod.rs

```diff
@@ -284,39 +284,64 @@ impl<R: RealNumberInternalTrait, E: IEnvironment<R>> Interpreter<R, E> {
         }
     }
 
-    fn apply_scheme_procedure(
+    fn init_procedure_env(
         formals: &Vec<String>,
         internal_definitions: &Vec<Definition>,
         expressions: &Vec<Expression>,
-        closure: Rc<RefCell<E>>,
+        env: &Rc<RefCell<E>>,
         args: ArgVec<R, E>,
     ) -> Result<Value<R, E>> {
-        let local_env = Rc::new(RefCell::new(E::new_child(closure.clone())));
         for (param, arg) in formals.iter().zip(args.into_iter()) {
-            local_env.borrow_mut().define(param.clone(), arg);
+            env.borrow_mut().define(param.clone(), arg);
         }
         for Definition(name, expr) in internal_definitions {
-            let value = Self::eval_expression(&expr, &local_env)?;
-            local_env.borrow_mut().define(name.clone(), value)
+            let value = Self::eval_expression(&expr, &env)?;
+            env.borrow_mut().define(name.clone(), value)
         }
         match expressions.split_last() {
             Some((last, before_last)) => {
                 for expr in before_last {
-                    Self::eval_expression(expr, &local_env)?;
+                    Self::eval_expression(expr, &env)?;
                 }
-                Self::eval_expression(last, &local_env)
+                Self::eval_expression(last, &env)
             }
             None => logic_error!("no expression in function body"),
         }
     }
 
+    fn apply_scheme_procedure(
+        formals: &Vec<String>,
+        internal_definitions: &Vec<Definition>,
+        expressions: &Vec<Expression>,
+        closure: Rc<RefCell<E>>,
+        args: ArgVec<R, E>,
+        is_tail: bool,
+    ) -> Result<Value<R, E>> {
+        let mut _local_env = None;
```

Comment:
variable should not begin with an underscore.
https://doc.rust-lang.org/1.0.0/style/style/naming/README.html

Reply:
And also, try add some tests?

Reply:
> 
> 
> variable should not begin with an underscore.
> https://doc.rust-lang.org/1.0.0/style/style/naming/README.html


```rust
        let mut local_env = None;
        let closure_ref = if is_tail {
            closure.borrow_mut().clear_definition();
            &closure
        } else {
            local_env = Some(Rc::new(RefCell::new(E::new_child(closure.clone()))));
            &local_env.as_ref().unwrap()
        };
```
If I write the code like this, rustc will warn:
warning: value assigned to `local_env` is never read
   --> src\interpreter\mod.rs:320:13
    |
320 |         let mut local_env = None;
    |             ^^^^^^^^^^^^^
    |
    = note: `#[warn(unused_assignments)]` on by default
    = help: maybe it is overwritten before being read?

warning: 1 warning emitted

Reply:
> 
> 
> And also, try add some tests?

Yeah, I do add a long test in parser::definition([parser.rs](https://github.com/Danielmelody/Ruschm/blob/feat-proper-tail-recursion/src/parser.rs#L732-L799)), and edit some old test to suit the change([parser.rs](https://github.com/Danielmelody/Ruschm/blob/feat-proper-tail-recursion/src/parser.rs#L720), [parser.rs](https://github.com/Danielmelody/Ruschm/blob/feat-proper-tail-recursion/src/parser.rs#L858), [parser.rs](https://github.com/Danielmelody/Ruschm/blob/feat-proper-tail-recursion/src/parser.rs#L896))